### PR TITLE
Fix local livehandler

### PR DESCRIPTION
**Pull request by andrii-suse**
While using developer mode locally, error message "unable to upgrade ws to command server" was popping up every second.
Debugging revealed that Controller was expecting `header->location` to be present in reply from WS, which wasn't the case:

```bless( { 'raw_size' => 194, 'max_message_size' => 2147483648, 'max_line_size' => 8192, 'content' => bless( { 'auto_relax' => 1, 'pre_buffer' => '', 'auto_decompress' => 1, 'body' => 2, 'headers' => bless( { 'cache' => [], 'headers' => { 'upgrade' => [ 'websocket' ], 'connection' => [ 'Upgrade' ], 'sec-websocket-accept' => [ 'jw5h2qP/usDGEUeXxVwzZ2KkOMI=' ], 'server' => [ 'Mojolicious (Perl)' ], 'date' => [ 'Wed, 06 May 2020 02:58:56 GMT' ] }, 'max_line_size' => 8192, 'state' => 'finished', 'max_lines' => 100 }, 'Mojo::Headers' ), 'buffer' => '', 'auto_upgrade' => 1, 'real_size' => 0, 'state' => 'finished', 'raw_size' => 160, 'read' => sub { "DUMMY" }, 'skip_body' => 1, 'events' => { 'read' => [ $VAR1->{'content'}{'read'} ] }, 'header_size' => 160 }, 'Mojo::Content::Single' ), 'events' => {}, 'finished' => 2, 'message' => 'Switching Protocols', 'code' => '101', 'state' => 'finished', 'version' => '1.1' }, 'Mojo::Message::Response' ) ```

This commit fixes the problem.

### CODE REVIEW

**Comment by okurz** [comment]
Good fix. @Martchus shouldn't tests fail when we have unhandled errors or is this one of the FIXMEs that we have in the test code?

File: lib/OpenQA/WebAPI/Controller/LiveViewHandler.pm [hunk 1]
```diff
@@ -386,13 +386,15 @@ sub connect_to_cmd_srv {
             # upgrade to ws connection if not already a websocket connection
             if (!$tx->is_websocket) {
                 my $location_header = ($tx->completed ? $tx->res->headers->location : undef);
-                if (!$location_header) {
+                if (!$location_header && !$tx->res->headers->upgrade) {
```
Comment (Martchus) [inline lib/OpenQA/WebAPI/Controller/LiveViewHandler.pm:389]: Not sure how costly `$tx->res->headers` is but I'd save the headers in a variable instead of calling it twice.

Comment (Martchus) [reply lib/OpenQA/WebAPI/Controller/LiveViewHandler.pm:389]: I'm also not sure whether this makes actually sense in all cases. Maybe we need to follow a normal redirection before we end up being able to upgrade? Of course it would make sense to limit the number of redirections to follow. Or does this Mojo for us.

---

@kraih Any suggestion or best practices for this? I actually just copied that code from our worker.

---

Note that the code I've copied from is (at this point) in `OpenQA::Worker::WebUIConnection::_setup_websocket_connection` and likely should be adjusted as well.

Comment (andrii-suse) [reply lib/OpenQA/WebAPI/Controller/LiveViewHandler.pm:389]: I generate wrappers, but those just start the services with custom OPENQA_BASE_PORT defined and all rest is default (i.e. no config file), i.e.:
`OPENQA_BASE_PORT=${port} src/script/openqa daemon`
`OPENQA_BASE_PORT=${port} src/script/openqa-livehandler daemon`
`src/script/worker --isotovideo 'os-autoinst/isotovideo' --host http://127.0.0.1:${port} --instance 1 --apikey 1234567890ABCDEF --apisecret 1234567890ABCDEF`
etc . Then use `job post`. 
Basically you can just execute this script and look into few-lines start.sh files to see that https://github.com/andrii-suse/environs/blob/master/.product/qa/branch/.example/fullstack.sh

Comment (kraih) [reply lib/OpenQA/WebAPI/Controller/LiveViewHandler.pm:389]: I don't understand this code at all. The whole `if (!$tx->is_websocket) { ` block makes no sense to me.

Comment (Amrysliu) [reply lib/OpenQA/WebAPI/Controller/LiveViewHandler.pm:389]: When I do not set `WORKER_HOSTNAME =` in `wokers.ini`,  the error message `unable to upgrade ws to command server` will be shown in webUI. I am not sure if this situation is the same with yours.

Comment (kraih) [reply lib/OpenQA/WebAPI/Controller/LiveViewHandler.pm:389]: fwiw. Redirecting WebSockets is an optional feature in RFC 6455, and Mojolicious does not support it.

Comment (Martchus) [reply lib/OpenQA/WebAPI/Controller/LiveViewHandler.pm:389]: @kraih So Mojolicious should handle this automatically and follow redirections as needed? I didn't write this block initially (just copied it from the worker) but I *suspect* the motivation behind this block is to catch the case when there's just a regular HTTP server reachable and no upgrade is possible (e.g. the web socket server is not running so one just gets the error message from the reverse proxy). I also remember having errors when calling `send()` because when the upgrade is not possible one just ends up with a regular tx and that case needs to be distinguished. The `if (!$tx->is_websocket) {` block prevents that.

Comment (andrii-suse) [reply lib/OpenQA/WebAPI/Controller/LiveViewHandler.pm:389]: @Amrysliu thanks you, that may be the case - I did provide `WORKER_HOSTNAME=127.0.0.1` and `WORKER_HOSTNAME=localhost` as parameter to `job post`, I shall try ini file as well.

Comment (Martchus) [reply lib/OpenQA/WebAPI/Controller/LiveViewHandler.pm:389]: @andrii-suse This settings *needs* to go into the worker config file. It is a worker configuration and not a job setting.

Comment (kraih) [reply lib/OpenQA/WebAPI/Controller/LiveViewHandler.pm:389]: But as it turns out 307/308 redirects work with Mojolicious by accident, so i [turned that into a feature](https://github.com/mojolicious/mojo/compare/8fc68f476968...375e86134e4c) upstream. And that means you can just set `$self->app->ua->max_redirects(10)` and let `Mojo::UserAgent` worry about redirects and limit enforcement.

**Comment by andrii-suse** [comment]
Thx to @Amrysliu I realised that the fix for this was actually different.
livehandler was trying to resolve `localhost.localdomain` and was failing with that.
I remembered that together with the commit I also added localhost.localdomain to /etc/hosts and now I see that old code started to work as well, so the commit isn't the fix.
Sorry for that

File: lib/OpenQA/WebAPI/Controller/LiveViewHandler.pm (hunk 1 above)
Comment (kraih) [reply lib/OpenQA/WebAPI/Controller/LiveViewHandler.pm:389]: Updated the comment above to clarify that all redirects are now supported for WebSockets with `Mojo::UserAgent`.
